src/libexpr/eval.cc

```diff
@@ -246,15 +246,37 @@ EvalState::EvalState(
     , repair(NoRepair)
     , emptyBindings(0)
     , rootFS(
-        settings.restrictEval || settings.pureEval
-        ? ref<SourceAccessor>(AllowListSourceAccessor::create(getFSSourceAccessor(), {},
-            [&settings](const CanonPath & path) -> RestrictedPathError {
-                auto modeInformation = settings.pureEval
-                    ? "in pure evaluation mode (use '--impure' to override)"
-                    : "in restricted mode";
-                throw RestrictedPathError("access to absolute path '%1%' is forbidden %2%", path, modeInformation);
-            }))
-        : getFSSourceAccessor())
+        ({
+            auto accessor = getFSSourceAccessor();
+
+            /* If we have a chroot store, make a union accessor to
+               make the chroot store available at its logical location
+               while still having the underlying directory
+               available. This is necessary for instance if we're
+               evaluating a file from the physical /nix/store while
+               using a chroot store. */
+            auto realStoreDir = dirOf(store->toRealPath(StorePath::dummy));
+            if (store->storeDir != realStoreDir) {
```

Comment:
I think we'll want this always
- for simplicity
- in order to expose chroot-store bugs sooner
- in preparation for other stores, e.g. an in-memory or write caching store wrapper for performance
- eventually for https://github.com/NixOS/nix/issues/9319

```suggestion
```

If we're not confident, I'd want a flag to make it unconditional so the team and any other volunteers can dogfood it.

Reply:
The test suite succeeds with this enabled unconditionally, but performance I don't think we should always enable it. For instance, every `maybeLstat()` call will now result in two `maybeLstat()` calls to the underlying accessors if the file doesn't exist. Though since `PosixSourceAccessor` has some caching it's probably not too bad.